com.unity.perception/Runtime/GroundTruth/Labeling/LabelManager.cs

```diff
@@ -0,0 +1,167 @@
+        LinkedHashSet<Labeling> m_LabelsPendingRegistration = new LinkedHashSet<Labeling>();
+        LinkedHashSet<Labeling> m_RegisteredLabels = new LinkedHashSet<Labeling>();
+
+        /// <summary>
+        /// Returns the set of registered Labeling components
+        /// </summary>
+        public IEnumerable<Labeling> registeredLabels => m_RegisteredLabels;
+
+        /// <summary>
+        /// Registers all pending labels.
+        /// Called once per frame during LateUpdate by the <see cref="PerceptionUpdater"/>.
+        /// </summary>
+        public void RegisterPendingLabels()
+        {
+            if (m_RegisteredLabels.Count == 0)
+                m_CurrentObjectIndex = k_StartingIndex;
+
+            foreach (var unregisteredLabel in m_LabelsPendingRegistration)
+            {
+                if (m_RegisteredLabels.Contains(unregisteredLabel))
+                    continue;
+
+                var instanceId = m_CurrentObjectIndex++;
+
+                RecursivelyInitializeGameObjects(
+                    unregisteredLabel.gameObject,
+                    new MaterialPropertyBlock(),
+                    unregisteredLabel,
+                    instanceId);
+
+                unregisteredLabel.SetInstanceId(instanceId);
+                m_RegisteredLabels.Add(unregisteredLabel);
+            }
+
+            m_LabelsPendingRegistration.Clear();
+        }
+
+        /// <summary>
+        /// Activates the given <see cref="IGroundTruthGenerator"/>.
+        /// <see cref="IGroundTruthGenerator.SetupMaterialProperties"/> will be called for all
+        /// <see cref="MeshRenderer"/> instances under each object containing a <see cref="Labeling"/> component.
+        /// </summary>
+        /// <param name="generator">The generator to register</param>
+        public void Activate(IGroundTruthGenerator generator)
+        {
+            m_ActiveGenerators.Add(generator);
+            foreach (var label in m_RegisteredLabels)
+                RecursivelyInitializeGameObjects(label.gameObject, new MaterialPropertyBlock(), label, label.instanceId);
+        }
+
+        /// <summary>
+        /// Deactivates the given <see cref="IGroundTruthGenerator"/>.
+        /// It will no longer receive calls when <see cref="Labeling"/> instances are created.
+        /// </summary>
+        /// <param name="generator">The generator to deactivate</param>
+        /// <returns>
+        /// True if the <see cref="generator"/> was successfully removed.
+        /// False if <see cref="generator"/> was not active.
+        /// </returns>
+        public bool Deactivate(IGroundTruthGenerator generator)
+        {
+            return m_ActiveGenerators.Remove(generator);
+        }
+
+        /// <summary>
+        /// Registers a labeling component
+        /// </summary>
+        /// <param name="labeledObject">the component to register</param>
+        public void Register(Labeling labeledObject)
```

Comment:
Why should this be public?

Reply:
Good catch, I'll change this to internal.